statistics/handle/globalstats/global_stats_async.go

```diff
@@ -0,0 +1,468 @@
+		if err1 != nil {
+			// no idx so idx = 0
+			err := a.dealWithSkipPartition(partitionID, isIndex, 0, err1)
+			if err != nil {
+				return err
+			}
+			if types.ErrPartitionStatsMissing.Equal(err1) {
+				continue
+			}
+		}
+		for idx, hist := range a.histIDs {
+			err1 := skipColumnPartition(sctx, partitionID, isIndex, hist)
+			if err1 != nil {
+				err := a.dealWithSkipPartition(partitionID, isIndex, idx, err1)
+				if err != nil {
+					return err
+				}
+				if types.ErrPartitionStatsMissing.Equal(err1) {
+					break
+				}
+			}
+			if idx == 0 {
+				realtimeCount, modifyCount, isNull, err := storage.StatsMetaCountAndModifyCount(sctx, partitionID)
+				if err != nil || isNull {
+					return err
+				}
+				// In a partition, we will only update globalStats.Count once.
+				a.globalStats.Count += realtimeCount
+				a.globalStats.ModifyCount += modifyCount
+			}
+		}
+	}
+	return nil
+}
+
+func (a *AsyncMergePartitionStats2GlobalStats) dealWithSkipPartition(partitionID int64, isIndex bool, idx int, err error) error {
+	switch {
+	case types.ErrPartitionStatsMissing.Equal(err):
+		return a.dealErrPartitionStatsMissing(partitionID)
+	case types.ErrPartitionColumnStatsMissing.Equal(err):
+		return a.dealErrPartitionColumnStatsMissing(isIndex, partitionID, idx)
+	default:
+		return err
+	}
+}
+
+func (a *AsyncMergePartitionStats2GlobalStats) dealErrPartitionStatsMissing(partitionID int64) error {
+	missingPart := fmt.Sprintf("partition `%s`", a.PartitionDefinition[partitionID].Name.L)
+	a.globalStats.MissingPartitionStats = append(a.globalStats.MissingPartitionStats, missingPart)
+	for _, histID := range a.histIDs {
+		a.skipPartition[skipItem{
+			histID:      histID,
+			partitionID: partitionID,
+		}] = struct{}{}
+	}
+	return nil
+}
+
+func (a *AsyncMergePartitionStats2GlobalStats) dealErrPartitionColumnStatsMissing(isIndex bool, partitionID int64, idx int) error {
+	var missingPart string
+	if isIndex {
+		missingPart = fmt.Sprintf("partition `%s` index `%s`", a.PartitionDefinition[partitionID].Name.L, a.tableInfo[partitionID].FindIndexNameByID(a.histIDs[idx]))
+	} else {
+		missingPart = fmt.Sprintf("partition `%s` column `%s`", a.PartitionDefinition[partitionID].Name.L, a.tableInfo[partitionID].FindColumnNameByID(a.histIDs[idx]))
+	}
+	if !a.skipMissingPartitionStats {
+		return types.ErrPartitionColumnStatsMissing.GenWithStackByArgs(fmt.Sprintf("table `%s` %s", a.tableInfo[partitionID].Name.L, missingPart))
+	}
+	a.globalStats.MissingPartitionStats = append(a.globalStats.MissingPartitionStats, missingPart)
+	a.skipPartition[skipItem{
+		histID:      a.histIDs[idx],
+		partitionID: partitionID,
+	}] = struct{}{}
+	return nil
+}
+
+func (a *AsyncMergePartitionStats2GlobalStats) ioWorker(sctx sessionctx.Context, isIndex bool) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			close(a.exitChan)
+			err = errors.New(fmt.Sprint(r))
+		}
+	}()
+	err = a.loadFmsketch(sctx, isIndex)
+	if err != nil {
+		close(a.exitChan)
```

Comment:
Better to add `close(a.exitChan)` in the defer func above instead of calling it each time when return.

Reply:
exitChan is to exit when meet the error. I think it should be renamed.